contrib/babelfishpg_tsql/src/pltsql_utils.c

```diff
@@ -889,3 +895,230 @@ get_pltsql_function_signature(PG_FUNCTION_ARGS)
+				BTEqualStrategyNumber, F_NAMEEQ,
+				CStringGetDatum(conname));
+
+	ScanKeyInit(&skey[1],
+				Anum_pg_constraint_connamespace,
+				BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(connamespace));
+
+	tgscan = systable_beginscan(tgrel, ConstraintNameNspIndexId,
+								true, NULL, 2, skey);
+
+	/* we are interested in the first row only */
+	if (HeapTupleIsValid(tuple = systable_getnext(tgscan)))
+	{
+		Form_pg_constraint con = (Form_pg_constraint) GETSTRUCT(tuple);
+		if (OidIsValid(con->oid))
+		{
+			result = con->oid;
+		}
+	}
+	systable_endscan(tgscan);
+	table_close(tgrel, AccessShareLock);
+	return result;
+}
+
+/*
+ * tsql_get_trigger_oid
+ *		Given name of a trigger, look up the OID.
+ *
+ * Returns InvalidOid if there is no such trigger.
+ */
+Oid
+tsql_get_trigger_oid(char *tgname)
+{
+	Relation		tgrel;
+	ScanKeyData 	key;
+	SysScanDesc 	tgscan;
+	HeapTuple		tuple;
+	Oid				result = InvalidOid;
+
+	/* search in pg_trigger by name */
+	tgrel = table_open(TriggerRelationId, AccessShareLock);
+	ScanKeyInit(&key,
+				Anum_pg_trigger_tgname,
+				BTEqualStrategyNumber, F_NAMEEQ,
+				CStringGetDatum(tgname));
+
+	tgscan = systable_beginscan(tgrel, TriggerRelidNameIndexId,
+								true, NULL, 1, &key);
+	
+	/* we are interested in the first row only */
+	if (HeapTupleIsValid(tuple = systable_getnext(tgscan)))
+	{
+		Form_pg_trigger pg_trigger = (Form_pg_trigger) GETSTRUCT(tuple);
+		if (OidIsValid(pg_trigger->oid))
+		{
+			result = pg_trigger->oid;
+		}
+	}
+	systable_endscan(tgscan);
+	table_close(tgrel, AccessShareLock);
+	return result;
+}
+
+/*
+ * tsql_get_proc_oid
+ *		Given name and namespace of a proc, look up the OID.
+ *
+ * Returns InvalidOid if there is no such proc.
+ */
+Oid
+tsql_get_proc_oid(char *proname, Oid pronamespace)
+{
+	HeapTuple		tuple;
+	CatCList		*catlist;
+	Oid				result = InvalidOid;
+
+	/* first search in pg_proc by name */
+	catlist = SearchSysCacheList1(PROCNAMEARGSNSP, CStringGetDatum(proname));
+	for (int i = 0; i < catlist->n_members; i++)
+	{
+		Form_pg_proc procform;
+		tuple = &catlist->members[i]->tuple;
+		procform = (Form_pg_proc) GETSTRUCT(tuple);
+		/* then consider only procs in specified namespace */
+		if (procform->pronamespace == pronamespace)
+		{
+			result = procform->oid;
+			break;
+		}
+	}
+	ReleaseSysCacheList(catlist);
+	return result;
+}
+
+/*
+ * tsql_get_temp_object_oid
+ *		Given relname of a temp object, look up the OID.
+ *		Search in list of ENRs registered in the current query environment by relname.
+ * Returns InvalidOid if there is no such proc.
+ */
+Oid
+tsql_get_temp_object_oid(char *relname)
+{
+	ListCell 		*lc;
+	char			*name;
+	int 			reloid;
+	Oid				result = InvalidOid;
+	List 			*enr_list = get_namedRelList();
+
+	foreach (lc, enr_list)
+	{	
+		reloid = ((EphemeralNamedRelationMetadata)lfirst(lc))->reliddesc;
+		name = ((EphemeralNamedRelationMetadata)lfirst(lc))->name;
+		if (!strcmp(name, relname))
+		{
+			result = reloid;
```

Comment:
Function `Datum get_enr_list()` unnecessarily creating result tuple for all rows, It will not be a good idea to use that when we are talking in terms of performance.

Reply:
Again, I am specifically talking about get_ENR() function. We cant use that?

Reply:
addressed in next commit